patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch

```diff
@@ -17,23 +17,189 @@
          {
              if (!this.field_70170_p.field_72995_K && (this.field_70963_i == null || this.field_70963_i.size() > 0))
              {
-@@ -540,6 +541,7 @@
+@@ -255,6 +256,8 @@
+     {
+         super.func_70014_b(p_70014_1_);
+         p_70014_1_.func_74768_a("Profession", this.func_70946_n());
++        String professionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(this.func_70946_n()).getName();
++        p_70014_1_.func_74778_a("professionName", professionName);
+         p_70014_1_.func_74768_a("Riches", this.field_70956_bz);
+         p_70014_1_.func_74768_a("Career", this.field_175563_bv);
+         p_70014_1_.func_74768_a("CareerLevel", this.field_175562_bw);
+@@ -283,14 +286,34 @@
+     public void func_70037_a(NBTTagCompound p_70037_1_)
+     {
+         super.func_70037_a(p_70037_1_);
+-        this.func_70938_b(p_70037_1_.func_74762_e("Profession"));
++        int professionId = p_70037_1_.func_74762_e("Profession");
++        String professionName = p_70037_1_.func_74779_i("professionName");
++        net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession profession = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForName(professionName);
++        String oldProfession = profession.getName();
++        if(oldProfession.equals(professionName) && professionId != profession.getId())
++        {
++            professionId = profession.getId();
++        }
++        else if(!professionName.isEmpty() && !oldProfession.equals(professionName))
++        {
++            professionId = 0;
++        }
++        this.func_70938_b(professionId);
+         this.field_70956_bz = p_70037_1_.func_74762_e("Riches");
+         this.field_175563_bv = p_70037_1_.func_74762_e("Career");
+         this.field_175562_bw = p_70037_1_.func_74762_e("CareerLevel");
+         this.field_175565_bs = p_70037_1_.func_74767_n("Willing");
+-
+-        if (p_70037_1_.func_150297_b("Offers", 10))
++        
++        String newProfessionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(func_70946_n()).getName();
++        if(professionId != func_70946_n()||!(professionName.equals(newProfessionName)||professionName.isEmpty()))
```

Comment:
I do not see where the potential NPE is in that area, the getProfessionForName method will never return null, if it was going to, it instead returns the value for farmer.  I will see about fixing the spacing.